### DISCUSSION TITLE
Let's talk about Appwrite's account preferences

### DISCUSSION THREAD

**Discussion by Meldiron** [appwrite/appwrite#2382] [Ideas]
![GitHub Topics (1)](https://user-images.githubusercontent.com/19310830/139842641-293ace50-cbe0-4b18-bc0c-dbbaeb9c0582.png)

Appwrite's account preferences let you store information on the user's objects to easily share them across multiple devices and sessions. I have user prefs improvement in my mind for a long time, so let's start a discussion! 🥳 My idea is to add multiple preferences for multiple permission levels:

- owner read, owner write (private, insecure) | For example `theme=dark`
- owner read, server write (private, secure) | For example: `discordUserId=4df65sf64s5d4ds9f15sf8d4sd468f`
- anyone read, owner write (public, insecure) | For example: `gender=Male`
- anyone read, server write (public, secure) | For example: `isVip=true`

---

Let's Chat! How can we improve Appwrite user preferences? What do you think about this initial idea? Can you see any problems or do you have an idea how to improve it even more?

**Comment by NeonSpork** [comment]
Ideas for extending the usefulness:

Currently there is a `user` that's on a per project basis, but it would be neat to explore making a `userProfile` with key/value pairs similar to documents. Although that begs the question why not just use the existing document structure to begin with 🤔

Perhaps it would be possible to extend the user object somehow, so that when you did `sdk.account.get()` it would return the baseline `user object` or override that object with an extended project-specific user if the project settings has one specified. One that has all the normal fields that the user object has, but _also_ additional fields as defined by the project maintainer.

**Reply by Meldiron** [reply]
I think I see what you are trying to do but I cannot understand the scenario when it is useful. Could you please describe a specific use case?

**Reply by NeonSpork** [reply]
One example I was thinking was in the event of a game, or a unity app, where other things such as `userName`, `optInNewsletter`, `highScore`, and other random things specific to the project could follow the users profile, rather than having to maintain a separate document collection with all those things for each individual user.

I'm imagining a scenario with appwrite as the backend for a web3 game or as the backend for a more traditional game, but I'm sure there are other non-game usecases where having a specific user profile would be useful.

Although, I'm not sure my idea is any better than using the existing document database for storing the items needed for such a profile anyway.

**Comment by eldadfux** [comment]
I think adding permissions to user preferences will add a lot of complexity and open questions for that kind of an implementation.

The nature of user preferences is always to be private for the user and insecure, meaning the user can change his own preferences (very good for metadata like theme, age, gender, etc...). Today, nobody can read the permissions beside the user or the project admin because if that was possible it was really easy for anyone to fetch the entire users list of a project.

For "secure" permissions the only admins/owners can write or read we had the idea of **users labels** that could only be set from the server API or using the console. I think this is better to achieve the different use-cases that could use this feature. 

The other major problem I think we have with user preferences is the lack of structure. This makes the future implementation of GraphQL or Response Models in the Appwrite SDKs very difficult.

I think we should probably allow to preset which keys are allowed, and accept only a specific type of file. Going with full DB collection permissions might be an option, but a bit more complex. The main advantage over just using the DB is the we can probably handle the results cache per user more effectively and really speed up this endpoint.

**Reply by bliyscom** [reply]
Custom fields like WordPress would be great for user accounts

**Comment by zdanl** [comment]
Has this been addressed? Clearly, non-user writeable account preferences, say account securityDetails, - or privileges per account preference as suggested by the thread author - would eliminate a lot of complexity like re-routing to a server side function to access the database. @Meldiron

**Reply by zdanl** [reply]
For ethereum (serverless...) wallet data and KYC data to suport building a NFT community & exchange on AppWrite.
